src/Evolution/Initialization/FunctionOfTimeFromFile.hpp

```diff
@@ -0,0 +1,145 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <memory>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
+#include "Domain/FunctionsOfTime/PiecewisePolynomial.hpp"
+#include "Domain/FunctionsOfTime/Tags.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Evolution/Initialization/Tags.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/Dat.hpp"
+#include "IO/H5/File.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Initialization {
+namespace Actions {
+/// \ingroup InitializationGroup
+/// \brief Initialize a `FunctionsOfTime` from dat files in an
+/// H5 file.
+///
+/// Uses:
+///  - DataBox:
+///    - `Initialization::Tags::FuncOfTimeFile`
+///    - `Initialization::Tags::FuncOfTimeSetNames`
+///
+/// DataBox changes:
+/// - Adds:
+///   - A `FunctionsOfTime` containing a `FunctionOfTime` for each dat
+///   file read from the H5 file
+///
+/// - Removes: nothing
+/// - Modifies: nothing
+struct FunctionOfTimeFromFile {
+  template <typename DataBox, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(DataBox& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& file_name{
+        db::get<Initialization::Tags::FuncOfTimeFile>(box)};
+
+    // Each element of dataset_names is a std::array<std::string, 2>, where
+    // element 0 is the dataset name in the h5 file, and element 1 is the
+    // name used in the std::unordered_map that will store the read-in functions
+    // of time
+    const auto& dataset_names{
```

Comment:
If you make the type explicit `std::array<std::string, 2>` then the comment can be shortened and we have the compiler checking more things

Reply:
Sure! Do you have a good feeling how I know when is a good idea to use `auto`? The explanation in the dev guide just says basically unless you have a `DataVector`, so I'd been defaulting to using auto when possible except for `DataVectors`